dexlib2/src/main/java/com/android/tools/smali/dexlib2/writer/DexWriter.java

```diff
@@ -1172,31 +1172,47 @@ private int writeDebugItem(@Nonnull DexDataWriter writer,
                 }
             }
         }
-        writer.writeUleb128(startingLineNumber);
 
-        writer.writeUleb128(parameterCount);
+        ByteArrayOutputStream tempByteOutput = new ByteArrayOutputStream();
+        DexDataWriter tempDataWriter = new DexDataWriter(tempByteOutput, 0, 64);
+        DebugWriter tempDebugWriter = new DebugWriter(stringSection, typeSection, tempDataWriter);
+
+        tempDataWriter.writeUleb128(startingLineNumber);
+
+        tempDataWriter.writeUleb128(parameterCount);
         if (parameterNames != null) {
             int index = 0;
             for (StringKey parameterName: parameterNames) {
                 if (index == parameterCount) {
                     break;
                 }
                 index++;
-                writer.writeUleb128(stringSection.getNullableItemIndex(parameterName) + 1);
+                tempDataWriter.writeUleb128(stringSection.getNullableItemIndex(parameterName) + 1);
             }
         }
 
         if (debugItems != null) {
-            debugWriter.reset(startingLineNumber);
+            tempDebugWriter.reset(startingLineNumber);
 
             for (DebugItem debugItem: debugItems) {
-                classSection.writeDebugItem(debugWriter, debugItem);
+                classSection.writeDebugItem(tempDebugWriter, debugItem);
             }
         }
         // write an END_SEQUENCE opcode, to end the debug item
-        writer.write(0);
-
-        return debugItemOffset;
+        tempDataWriter.write(0);
+
+        tempDataWriter.flush();
+        byte[] debugInfo = tempByteOutput.toByteArray();
+        ByteBuffer wrapBytes = ByteBuffer.wrap(debugInfo);
+        int cacheBytes = debugInfoCaches.getOrDefault(wrapBytes, -1);
```

Comment:
I am not sure how `hashCode` for a `ByteBuffer` is implemented, but most likely it will hash over all bytes. It might give a large overhead having to calculate a hash over all bytes for large byte arrays. How about wrapping the byte array in a different object with an custom `hashCode` and `equals` with the hash code calculated from the byte array length and some fixed number of bytes in byte array. Maybe look at the generated debug info to see where the most entropy is.

Reply:
1. Yes, after compiling the APK, we employ a certain strategy to rewrite the DEX. This optimization has been applied to our app for over six months and has proven to be very stable.

2. For DEX files generated by higher versions of the Android Gradle Plugin (AGP), this optimization can reduce the size by approximately 7/24.

3. Following your suggestion, I extracted a `DebugInfoCache` class. However, the `hashCode` method still iterates over the entire array. Nevertheless, we've implemented a layer of caching for the `hashCode` computation. Generally speaking, `DebugInfo` is not very large, and compared to the importance of accuracy, the cost of computing the `hashCode` is minimal.